File: 卡码/代码随想录/1/09-动态规划/Python/02-背包问题/01-01背包/LeetCode_1049.py
#1049. 最后一块石头的重量 II
from typing import *

class Solution:
  def lastStoneWeightII(self, stones: List[int]) -> int:
    total:int=sum(stones)
    target:int=total//2
    dp:List[int]=[0]*1501
    for i in range(len(stones)):
      for j in range(target,stones[i]-1,-1):
        dp[j]=max(dp[j],dp[j-stones[i]]+stones[i])
    return total-2*dp[target]

File: 卡码/代码随想录/1/09-动态规划/Js/02-背包问题/01-01背包/LeetCode_1049.js
//1049. 最后一块石头的重量 II
const lastStoneWeightII = function(stones) {
	const sum=stones.reduce((prev,cur)=>prev+cur,0);
	const target=Math.floor(sum/2);
	const dp=new Array(1501).fill(0);
	for(let i=0;i<stones.length;i++){
		for(let j=target;j>=stones[i];j--){
			dp[j]=Math.max(dp[j],dp[j-stones[i]]+stones[i]);
		}
	}
	return sum-2*dp[target];
};